web/controllers/api/key_controller.ex

```diff
@@ -25,6 +25,7 @@ defmodule HexWeb.API.KeyController do
   def create(conn, params) do
     case Key.create(conn.assigns.user, conn.params) |> HexWeb.Repo.insert do
       {:ok, key} ->
+        HexWeb.AuditLog.create(conn.assigns.user, "key.generate", key) |> HexWeb.Repo.insert!
```

Comment:
@ericmj I wrapped most of the DB operations and audits within the same transaction but here I encountered an interesting problem. We pattern match the return value of Repo.insert to gracefully handle validation errors, but when this operation would be wrapped in a transaction, `Repo.transaction(fun)` would either return `{:ok, result_of_fun}` or `{:error, :rollback}` in which keys the changeset.errors are somewhat lost. Can you think of any elegant solution? I tried stuff like this: https://gist.github.com/wojtekmach/84a81416b39815de93f3, but it led me nowhere.

Similar problem here: https://github.com/wojtekmach/hex_web/blob/audit_logs/web/controllers/api/release_controller.ex#L112:L114